compiler/hash-typecheck/src/diagnostics/macros.rs

```diff
@@ -0,0 +1,137 @@
+//! Utility macros for performing various operations when it comes to
+//! working with reports
+
+const TC_FATAL_ERROR_MESSAGE: &str =
+    "A fatal error occurred during typechecking on the reported node";
+
+/// The macro `tc_panic` is essentially used to create a [Report], build
+/// it and then print it whilst panicking. This is a useful utility to denote
+/// where a panic occurs and provide additional context about where the panic
+/// occurred in regards to traversing the sources.
+///
+/// The `storage` argument to the macro is used to access the storage in order
+/// to resolve the [SourceMap] and [GlobalStorage] so that the term location can
+/// be fetched.
+pub macro tc_panic {
+    ($term: expr, $storage:expr, $fmt: expr, $($arg:tt)*) => {
+        {
+            use crate::storage::AccessToStorage;
+            let storages = $storage.storages();
+
+            // get the sources and and the location from the term
+            let term_location = storages.location_store().get_location($term);
+            let sources = storages.source_map();
+
+            // format the message and build the report
+            let message = format!($fmt, $($arg)*);
+            let mut report = hash_reporting::builder::ReportBuilder::new();
+            report
+                .with_kind(hash_reporting::report::ReportKind::Internal)
+                .with_message(message);
+
+            if let Some(location) = term_location {
+                builder
+                    .add_element(hash_reporting::report::ReportElement::CodeBlock(hash_reporting::report::ReportCodeBlock::new(
+                        location,
+                        "here",
+                    )));
+            }
+
+            println!("{}", hash_reporting::writer::ReportWriter::new(report.build(), sources));
```

Comment:
See stderr comment above

Reply:
Also, shouldn't this also print the term itself?

Reply:
Should this always happen, or should it be controlled by the formatting arguments?

Reply:
I think it should always happen, the user of the macro can always print more terms if they want, but it should print the "subject" term of the panic i think (just like it prints its location).